src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt

```diff
@@ -211,27 +226,68 @@ internal class ModelCheckingStrategy(
         }
     }
 
+    private inner class CrashChoosingNode : InterleavingTreeNode() {
+        override fun nextInterleaving(interleavingBuilder: InterleavingBuilder): Interleaving {
+            val isLeaf = maxNumberOfSwitches == interleavingBuilder.numberOfCrashes + interleavingBuilder.numberOfSwitches
+            if (isLeaf) {
+                finishExploration()
+                if (!isInitialized)
+                    interleavingBuilder.addLastNoninitializedNode(this)
+                return interleavingBuilder.build()
+            }
+            val choice = chooseUnexploredNode()
+            interleavingBuilder.addCrashPosition(choice.value)
+            val interleaving = choice.node.nextInterleaving(interleavingBuilder)
+            updateExplorationStatistics()
+            return interleaving
+        }
+    }
+
+    private inner class SwitchOrCrashChoosingNode : InterleavingTreeNode() {
+        init {
+            choices = listOf(Choice(SwitchChoosingNode(), 0), Choice(CrashChoosingNode(), 1))
+        }
+
+        override fun nextInterleaving(interleavingBuilder: InterleavingBuilder): Interleaving {
+            val child = chooseUnexploredNode()
+            interleavingBuilder.addCrashOrSwitchChoice(child.value)
+            val interleaving = child.node.nextInterleaving(interleavingBuilder)
+            updateExplorationStatistics()
+            return interleaving
+        }
+    }
+
     private inner class Choice(val node: InterleavingTreeNode, val value: Int)
 
     /**
      * This class specifies an interleaving that is re-producible.
      */
     private inner class Interleaving(
         private val switchPositions: List<Int>,
+        private val crashPositions: List<Int>,
         private val threadSwitchChoices: List<Int>,
-        private var lastNotInitializedNode: SwitchChoosingNode?
+        private val switchOrCrashChoices: List<Int>,
+        private var lastNotInitializedNode: InterleavingTreeNode?
     ) {
         private lateinit var interleavingFinishingRandom: Random
         private lateinit var nextThreadToSwitch: Iterator<Int>
+        private lateinit var nextCrashSwitchChoice: Iterator<Int>
+        private var currentSwitchCrashChoice: Int = -1
```

Comment:
I guess, this logic can be simplified.
You can check `isSwitchPosition` and `isCrashPosition` without comparing event type choices, because every execution position implicitly corresponds to a possible crash or switch. Every time `isSwitchPosition` or `isCrashPosition` is called, the execution position is increased, so there are no collisions.